[Table of Contents](../../index.ipynb)

# FRC Analytics with Python - Session 10
# File Operations and Comprehensions
**Last Updated: 6 October 2021**  

Honestly, file operations and comprehensions don't have much in common with each other. They are combined here because they fit together nicely in one session.

##### If Using Google Colab
It's best if you clone the *pyclass_frc* Github repo and run this notebook from your local computer. But if you would like to run it from Google Colab, uncomment and run the two lines in the next cell. (*Don't delete the exclamation points at the start of the line!*) They will download several data files into the local folder on Google Colab. The files are needed to run the examples and do the exercises in this notebook.

In [ ]:
# !wget -nv https://raw.githubusercontent.com/irs1318dev/pyclass_frc/master/sessions/s09_files_and_comprehensions/get_files_s9.sh
# !bash get_files_s9.sh

## I. File Operations

### A. Reading Data From a Text File
I'm sure you all understand why the ability to read or write to files is essential. So we'll just jump right in.

Python contains a built-in function called `open()`, which we'll use to work with files. Let's open and read data from our first file.

In [1]:
# Opening a file and displaying contents
txtfile = open(r"hopper.txt", "rt")     # Line 1
txt = txtfile.read()                    # Line 2
txtfile.close()                         # Line 3
print(txt[:454])                        # Line 4

The Wit and Wisdom of Grace Hopper
"Life was simple before World War II. After that, we had systems."
That observation comes from one who was present at the creation of the age of systems -- Rear Admiral Grace Hopper (US Navy, Retired), who spoke on the campus of the Ohio State University, Columbus, on Feb. 5, 1987, as part of a year-long celebration of the twentieth anniversary of the formation of the Department of Computer and Information Science.



#### 1. Line 1 - The `open()` Function

##### a. First Parameter - File Path
First, in line 1, we called the `open()` function and the first paramter was a file name, `"hopper.txt"`. As written, with just the file name, the `open()` function will expect the *hopper.txt* file to be located in the same folder as this notebook. If *hopper.txt* were located in some other folder, `open()` would throw an error. There are several ways to tell `open()` to look in different folders.

* **Relative Path to Subfolder:** Referring to files in subfolders of the current folder is easy. Just prepend the folder name and a forward slash to the file name. For example, if *hopper.txt* was in a subfolder called *text_files*, we would pass this value to `open()`: *`"text_files/hopper.txt"`*
* **Relative Path to Parent Folder:** Adding two periods (*`..`*) to the path tells Python to go to up one folder. For example, if *hopper.txt* was located in the parent folder of this notebook, we would pass *`"../hopper.txt"`* to `open()`. If *hopper.txt* was two folders up, we would pass *`"../../hopper.txt"`*.
* **Relative Paths to Sibling Folders:** The techniques in the previous two examples can be combined to get to a sibling folder. For example, *`"../brilliant_computer_scientists/hopper.txt"`* will jump up to a parent folder and then down to a sibling folder called *brilliant_computer_scientists*.
* **Absolute Paths:** An absolute path, starting with the root folder, can refer to any file on the computer. For example, *`"C:/Users/stacy/Projects/pyclass_frc/sessions/s10_comprehensions_and_files/hopper.txt"`* is an absolute path on a Windows computer.  On a linux system the path might look like *`"/home/Users/stacy/Projects/pyclass_frc/sessions/s10_comprehensions_and_files/hopper.txt"`* and on an apple computer it might look like *`"/Users/stacy/Projects/pyclass_frc/sessions/s10_comprehensions_and_files/hopper.txt"`*

Avoid absolute paths. If we used the path *`"C:/Users/stacy/Projects/pyclass_frc/sessions/s10_comprehensions_and_files/hopper.txt"`* to get to the *hopper.txt* file, then the code would only work for users named Stacy who are using a Windows machine. It's better to use relative paths.

##### b. Changing the Working Directory
An alternative to passing long file paths to the `open()` function is to change the working directory. Unless you specify otherwise, Python will assume all files are located in the working directory. Let's display our current working directory for this Jupyter notebook.

In [2]:
# Display the current working directory (CWD)
import os
def_cwd = os.getcwd()
def_cwd

'C:\\Users\\pperv\\Desktop\\Robotics\\Robotics Training\\pyclass_frc\\sessions\\s09_files_and_comprehensions'

The `getcwd()` function in the `os` module displays the current working directory.
By default, the working directory for a Jupyter notebook is the folder that the notebook file (*.ipynb* file) is stored in.

In [3]:
# Change the CWD
os.chdir("misc_files")
os.getcwd()

'C:\\Users\\pperv\\Desktop\\Robotics\\Robotics Training\\pyclass_frc\\sessions\\s09_files_and_comprehensions\\misc_files'

Our current working directory is now the *misc_files* folder. Let's change back to our original working directory.

In [4]:
# Restore the original CWD
os.chdir(def_cwd)
os.getcwd()

'C:\\Users\\pperv\\Desktop\\Robotics\\Robotics Training\\pyclass_frc\\sessions\\s09_files_and_comprehensions'

##### c. Working Directory for Modules Run from Command Line
The situtation is a little different when running Python modules from a command line. To demonstrate, we've placed a simple python module, *cwd.py* in the *misc_files* folder. The *cwd.py* file prints it's current working directory. Here are it's contents:

```python
import os

print()
print("Current Working Directory")
print(os.getcwd())
```

Let's see what the CWD is when we run *cwd.py* from a parent folder. We can execute CLI commands from a *Jupyter* code cell by starting the line with `!`. On MS Windows, Jupyter uses the command prompt for CLI commands. If you've downloaded the course and are running it on your local computer, run the first cell. If you are using Google Colab, Mac, or Linux, run the second cell.

In [5]:
# First, verify that cwd.py is located in the misc_files subfolder.
!dir misc_files /B

cwd.py
districts.json
matches.json
temp.json
TTY33ASR.jpg
YaleNews_hopper-grace.UNIVAC.102635875-CC_0.jpg


In [ ]:
# Run this cell if using Mac, Google Colab, or Linux
!ls

In [7]:
# Run cwd.py from a parent folder
!python misc_files/cwd.py


Current Working Directory
C:\Users\pperv\Desktop\Robotics\Robotics Training\pyclass_frc\sessions\s09_files_and_comprehensions


Interesting! The *cwd.py* file is located in the *misc_files* subfolder, but the current directory is not set to *misc_files*. It's set to the parent folder from which we *ran* *cwd.py*. Now let's run *cwd.py* from the *misc_files* subfolder.

In [8]:
# Run cwd.py from it's own folder
# Note that you can cram two commands onto one line with '&&'.
!cd misc_files && python cwd.py


Current Working Directory
C:\Users\pperv\Desktop\Robotics\Robotics Training\pyclass_frc\sessions\s09_files_and_comprehensions\misc_files


When we first changed directories to *misc_files* and then ran *cwd.py* from that directory, Python's current working directory was set to *misc_files*. 

The bottom line is that when running a Python module from a command line, the current working diretory will initially be set to the *folder from which we ran the module*, not the folder the module is located in. These two folders will often be the same folder, but not always.

##### d. Second Parameter - The Mode Parameter
The second parameter, `"rt"` specifies the mode. The first letter, `"r"`, specifies that we want to open the file for reading data. We can write data to a file by replacing `"r"` with `"w"`. If the file already exists, using `"w"` will overwrite the file. The value `"x"` can also be used to write data to a file. It will not overwrite an existing file, but will throw an error if the file already exists.

The second character, `"t"`, is used for text files. For binary files, we would use `"b"`.

##### e. `Open()` Function's Return Value
The `Open()` does not return the contents of the file, at least not directly. It returns a file object, which we saved to a variable named `txtfile`.

[The full documentation for `open()` is here.](https://docs.python.org/3/library/functions.html#open)

#### 2. Line 2 - Getting the Content of the File
In line 2, we we called the `read()` method on the `txtfile` object. The `read()` method returned the entire contents of the text file as a string. We could also have used `readlines()`, which returns a list, where each list item is one row from the file. Python splits the file into rows using the newline character, '/n'.

In [9]:
# Opening a file and using 
txtfile = open(r"hopper.txt", "rt")  # Line 1
txtlist = txtfile.readlines()        # Line 2
txtfile.close()                      # Line 3
print(txtlist[18])                   # Line 4 - Famous Quote from Grace Hopper

On the building of bigger computers: "In pioneer days they used oxen for heavy pulling, and when one ox couldn't budge a log, they didn't try to grow a larger ox. We shouldn't be trying for bigger computers, but for more systems of computers."



The `readlines()` method makes it easy to display just a single row from the text file. In our example we displayed the 19th line (remember, the first line has an index of 0), which contains a famous quote from
Grace Hopper.

#### 3. Line 3 - Closing the File
We called the `.close()` method in line 3 to close the file object. Methods like `read()` or `readlines()` won't work once the file is closed. In general, for small programs, nothing bad is going to happen if you don't explicitly close the file with `.close()`. Python will close the file automatically once the program or enclosing function exits. Still, bad things have been known to happen when programmers didn't close their files, so always close files when you are done with them.

### B. Using the `with` Statement
Programmers don't like to have to remember to do things like closing files. You can use a `with` statement to ensure files are closed when you are finished working with them. Here is an example:

In [10]:
# Opening a file and using 
with open(r"hopper.txt", "rt") as txtfile:  # Line 1
    txtlist = txtfile.readlines()           # Line 2
print(txtlist[15])                          # Line 3

In 1951 she discovered the first computer "bug." It was a real moth, which she pasted into the UNIVAC I logbook. In 1952 she had an operational compiler. "Nobody believed that," she said. "I had a running compiler and nobody would touch it. They told me computers could only do arithmetic." Admiral Hopper is also the "progenitor" of COBOL, which she was working on in 1955. In 1967, she was recalled to the Navy and served with the Naval Data Automation Command until she retired. Her mission was to preside over the Navy's efforts to maintain uniformity in computer languages. In 1983 she earned a special Presidential appointment to flag rank as admiral. She is now a consultant for Digital Equipment Corporation.



The `with` statement is followed by an indented block of code. We work with the file object within the indented block. The file object will automatically be closed when we exit the indented block in line 3. Another benefit of using `with` is that if an error occurs in the indented block of code, Python will ensure that the file gets closed anyway. Using `with` is  a best practice for working with files in Python.

The `with` statement can be used for things other than files. It is intended to be used with special objects called context managers. You can learn more than you probably want to know about context managers [at this link](https://docs.python.org/3/reference/datamodel.html#context-managers).

### C. End of the Line
If you were reading closely, then you noticed that we mentioned something about a *newline* character. A newline character is a special character used by Python to mark the end of a row in a text file, or to tell Python to move to a new row when displaying text. It can occur in a string just like any other character. In Python and many other programming languages, you can insert a newline character into a string using a backslash followed by an 'n', like so: `\n`.

In [11]:
multi_line_string = "This is line 1.\nThis is line 2."
multi_line_string

'This is line 1.\nThis is line 2.'

So far, the results are not very impressive. The notebook is just displaying the newline character in the string. The outcome is better if we pass the string to the `print()` function.

In [12]:
print(multi_line_string)

This is line 1.
This is line 2.


You may be thinking that `\n` looks like *two* characters, not one. That's understandable, but Python's `ord()` function shows that `\n` maps to a single character code.

In [14]:
print("Character code for newline:", ord("\n"))
# The next row uses a list comprehension, which will be covered later in
#   this session.
print("For comparison, character codes for IRS:", [ord(c) for c in "IRS"])

Character code for newline: 10
For comparison, character codes for IRS: [73, 82, 83]


We use the backslash notation because our alphabet doesn't have a symbol that means newline. "\n" is sometimes referred to as a control character, or a non-printable character, or an escape sequence character. There are other control characters in Python:

In [15]:
print("Include a backslash (\\) in a string with '\\\\'.")
print("Include quotation marks (\"\") in a string with '\\\"'.")
print("Include a tab character (\t) in a string with '\\t'.")

Include a backslash (\) in a string with '\\'.
Include quotation marks ("") in a string with '\"'.
Include a tab character (	) in a string with '\t'.


Technically and historically, the '\n' character is called a line feed character. Linux and older Unix operating systems use this character by default to mark the end of lines in text files. Microsoft Windows, on the other hand, uses two characters, a carriage return character ('\r') and a line feed character ('\n') to mark the end of a line. That's why, when you push files to a Github repository from Windows, you see all those messages about the end of line characters on the screen. Git is automatically converting your text files from using Windows style "\r\n" sequences to Linux style "\n" sequences. And if it's not confusing enough that Windows and Linux use different characters to mark the end of a line, consider that old Apple Macintosh operating systems used just a carriage return ('\r') to mark the end of a line.

**Warning: Historical Stuff that is Interesting to your Mentor and Probably No One Else!**  
How did it get to be this way? Video monitors were nonexistant or uncommon before the late 1970s. Anyone who wanted to interact in real time with a computer used a teleprinter, like the one pictured below.
![Vintage Teletype](misc_files/TTY33ASR.jpg)

The user would type commands on the teletype, seeing what they typed on a sheet of paper. They would hit a key to send the command to a computer, and the output from the command would print out on the paper. The command line interface that we are currently using is an on-screen version of the old teletype interface. The print head would move left to right across the page and the paper would be moved up and down by the roller. To start typing a new line, the user would hit the *RE-TURN* key to move the print head back to the left side of the page, and the *LINE FEED* key to move the paper up one row, effectively moving the print head down to the beginning of the next line.

So why does Windows use '\r\n' line endings while Linux uses just '\n'? Microsoft Windows uses '\r\n' because Microsoft's earlier operating system, MS-DOS, used '\r\n'. MS-DOS used '\r\n' because back in about 1980 when MS-DOS was being developed, they wanted MS-DOS to be compatible with another popular micro-computer operating system called CP/M, which had been developed in 1974 at a company called Digital Equipment Corporation. CP/M used '\r\n' because that sequence worked better for sending text to teletype machines like the ASR-33 teletype pictured above.

Linux doesn't use '\r\n' because Linux was designed to be compatible with an older operating system called Unix (Bell Labs, 1969), which itself was inspired by an even older operating system called MULTICS (Cooperattive project of MIT, Bel Labs, and General Electric, 1964). MULTICS used a device driver to communicate with the teletype (very innovative at the time). So text files could contain the simpler '\n' line ending and the device driver would convert '\n' to whatever sequence of characters the teletype needed to advance to the next line.

So there you have it. Windows uses '\r\n' because the developers of CP/M  in 1974 didn't use a device driver to talk to teletype machines. This might have been due to CP/M being a single-tasking operating system designed to work on small computers with 8-bit processors and no more than 64 kilobytes of memory. I'm guessing there was not a lot of extra bandwidth to be running a device driver. MULTICS, on the other hand, was designed to run on more powerful mainframe computers, like the GE-600 series.

**And Now Back to Modern Times**  
Most of the tools used by our robotics team take care of newline sequences without much trouble, so there's no need to memorize all of the different end of line sequences. But differences between newline sequences have caused headaches for many people in the past and could do so again in the future, so it's important to understand that different operating systems use different sequences. For Python, just use '\n'.

### D. Writing to a Text File
Here is an example of writing text to a file:

In [19]:
# Writing Text to a File
with open("numfile.txt", "wt") as numfile:
    for num in range(5):
        line = num * " " + str(num) + "\n"
        numfile.write(line)
    for num in range(5, -1, -1):
        line = num * " " + str(num) + "\n"
        numfile.write(line)    

The code for writing to a file looks almost exactly the same as the code for reading a file. There are only a couple differences.
* We just replaced "r" with "w" in the second parameter of the `open()` method.
* We used the file object's `.write()` method. Note that we have to manually insert a newline character at the end of each line.

We could use the `print()` function instead of `.write()`.

In [18]:
# Writing print()
with open("numfile.txt", "wt") as numfile:
    for num in range(5):
        print(num * " " + str(num), file=numfile)
    for num in range(5, -1, -1):
        print(num * " " + str(num), file=numfile)

We didn't have to manually add a newline character since `print()` appends a newline character to the output by default. We did have to pass the file object to the `file` parameter. Let's see what the file looks like.

In [22]:
with open("numfile.txt", "rt") as numfile:
    for line in numfile:
        print(line, end="")

0
 1
  2
   3
    4
     5
    4
   3
  2
 1
0


The code sample above shows that for text files, we can iterate over each line in the file. Note that becase each lin already contains a newline character, we used the `print()` function's `end` parameter to disable the newline at the end of each line. This keeps the output from being double-spaced.

## II. JSON Files
JSON, which stands for Javascript Object Notation, refers to a technique for exchanging and storing structured data as text. As you might guess from the name, JSON is derived from Javascript, but it is now used in many systems and programming languages. Let's take a look at a simple JSON file.

In [23]:
# Open a text file that contains information formatted as JSON
with open("misc_files/districts.json", "rt") as jfile:
    for lin_num, line in enumerate(jfile):
        print(line, end="")
        if lin_num > 11:
            break

[
 {
  "abbreviation": "chs", 
  "display_name": "FIRST Chesapeake", 
  "key": "2020chs", 
  "year": 2020
 }, 
 {
  "abbreviation": "fim", 
  "display_name": "FIRST In Michigan", 
  "key": "2020fim", 
  "year": 2020
 }, 


Only the first dozen or so lines are displayed to save space. The *districts.json* file contains information about all FIRST competition districts. Unlike the information in the *hopper.txt* file, which contains freeform text, the contents of *districts.json* must comply with a strict format.
* Sequences of items are enclosed in square brackets, like a Python list.
* Groups of key-value pairs are enclosed in curly braces, like a Python dictionary.
* Items are separated by commas
* Key-value pairs are separated by colons.
* All text is enclosed in quotes.

The *hopper.txt* file contains unstructured data, whereas the *districts.json* file contains structured data.

Python contains a handy JSON module that makes it easier to work with JSON data.

In [24]:
# Opening and Exploring a JSON file
import json

with open("misc_files/districts.json", "rt") as jfile:
    districts = json.load(jfile)
    
print("Type of object returned by `json.load()`:", type(districts))
print("Length of `districts` list:", len(districts))
print("Object type for each element of `districts` list:", type(districts[0]))
print()
print("First three elements of `districts` list:")
districts[1:4]

Type of object returned by `json.load()`: <class 'list'>
Length of `districts` list: 11
Object type for each element of `districts` list: <class 'dict'>

First three elements of `districts` list:


[{'abbreviation': 'fim',
  'display_name': 'FIRST In Michigan',
  'key': '2020fim',
  'year': 2020},
 {'abbreviation': 'fma',
  'display_name': 'FIRST Mid-Atlantic',
  'key': '2020fma',
  'year': 2020},
 {'abbreviation': 'fnc',
  'display_name': 'FIRST North Carolina',
  'key': '2020fnc',
  'year': 2020}]

If we use the `json.load()` method to read data from the file object, the JSON data is converted to a Python list of dictionary objects. We can save data to a JSON file in a similar fashion using `json.dump()`.

In [25]:
us_temps = {"description":
    {"title": "Contiguous U.S., Average Temperature, January-December",
        "units": "Degrees Fahrenheit",
        "base_period": "1901-2000",
        "missing": "-99"},
    "data": {
        "194012": {
            "value": "51.89",
            "anomaly": "-0.13"},
        "196012": {
            "value": "51.44",
            "anomaly": "-0.58"},
        "198012": {
            "value": "52.39",
            "anomaly": "0.37"},
        "200012": {
            "value": "53.27",
            "anomaly": "1.25"},
        "201612": {
            "value": "54.92",
            "anomaly": "2.90"}}
}
with open("temps.json", "wt") as tempfile:
    json.dump(us_temps, tempfile)

## III. Comprehensions
Now we'll take a break from files and review list and dictionary comprehensions. Comprehensions are a compact syntax for generating lists and dictionaries. They can be faster than using for loops in some circumstances, but we probably won't see any difference with relatively small datasets we use for FRC. Still, the compact syntax is quite neat. Most Python devotees would consider it good form to use a comprehension in place of a for loop.

Remember the `districts` JSON data object we used earlier? Suppose we wanted to build a list that contains just the district abbreviations. We could do that with a traditional `for` loop.

### A. Lists
Remember the `districts` JSON data object we used earlier? Suppose we wanted to build a list that contains just the district abbreviations. We could do that with a traditional `for` loop.

In [4]:
# Quick Refresher on districts data
districts[:3]

NameError: name 'districts' is not defined

In [27]:
# A for loop to build a list of district abbreviations.
codes = []
for district in districts:
    codes.append(district["abbreviation"])
codes

['chs', 'fim', 'fma', 'fnc', 'in', 'isr', 'ne', 'ont', 'pch', 'pnw', 'tx']

With a list comprehension, we can build the list in just one line of code.

In [28]:
# List comprehension that buils a list of abbreviations
codes = [district["abbreviation"] for district in districts]
codes

['chs', 'fim', 'fma', 'fnc', 'in', 'isr', 'ne', 'ont', 'pch', 'pnw', 'tx']

We've basically compressed the *for* loop into one line and put it inside the list's square brackets.

We can also do filtering in a list comprehension. Suppose we only are interested in district abbreviations that start with 'f':

In [10]:
# List comprehension with filtering
[district["abbreviation"] for district in districts if district["abbreviation"][0] == "f"]

NameError: name 'districts' is not defined

### B. Dictionaries
We can also use comprehensions to generate dictionaries:

In [8]:
# Dictionary of tuples with squares
tuple_squares = {str(x): (x, x**2) for x in range(10) if x % 3 != 0}
tuple_squares

{'1': (1, 1),
 '2': (2, 4),
 '4': (4, 16),
 '5': (5, 25),
 '7': (7, 49),
 '8': (8, 64)}

To create a dictionary comprehension, we drafted expressions for both the dictionaries key and value, separated by a colon. The key and value expressions are evaluated for every output of the iterator following the `for` keyword. The `if` expression filters out values where `x` is equal to 3. Finally, the entire expression is placed in curly braces.

### C. More Comprehension Examples
The *matches.json* file contains the complete match results from the 2020 FIRST Robotics district competition at Glacier Peak High School in Snohomish, WA.

In [244]:
import json
with open("misc_files/matches.json", "rt") as jfile:
    matches = json.load(jfile)

In [245]:
print("Object type of matches data:", type(matches))
print("Length of matches list:", len(matches))
print()
print("First match in list:")
matches[0]

Object type of matches data: <class 'list'>
Length of matches list: 89

First match in list:


{'actual_time': 1583104716,
 'alliances': {'blue': {'dq_team_keys': [],
   'score': 136,
   'surrogate_team_keys': [],
   'team_keys': ['frc2930', 'frc2976', 'frc4918']},
  'red': {'dq_team_keys': [],
   'score': 148,
   'surrogate_team_keys': [],
   'team_keys': ['frc4911', 'frc2910', 'frc4173']}},
 'comp_level': 'f',
 'event_key': '2020wasno',
 'key': '2020wasno_f1m1',
 'match_number': 1,
 'post_result_time': 1583105046,
 'predicted_time': 1583104822,
 'score_breakdown': {'blue': {'adjustPoints': 0,
   'autoCellPoints': 40,
   'autoCellsBottom': 0,
   'autoCellsInner': 2,
   'autoCellsOuter': 7,
   'autoInitLinePoints': 15,
   'autoPoints': 55,
   'controlPanelPoints': 0,
   'endgamePoints': 55,
   'endgameRobot1': 'Hang',
   'endgameRobot2': 'Hang',
   'endgameRobot3': 'Park',
   'endgameRungIsLevel': 'NotLevel',
   'foulCount': 0,
   'foulPoints': 0,
   'initLineRobot1': 'Exited',
   'initLineRobot2': 'Exited',
   'initLineRobot3': 'Exited',
   'rp': 0,
   'shieldEnergizedRankingPo

#### 1. Understanding the Match Data
The matches JSON data contains information on all of the matches, includng both qualification and playoffs. The *comp_level* key-value pair tells us whether the match was a qualification or playoff match. We can use a list comprehension to get a list of all the *comp_level* codes.

In [72]:
# List of all match competition levels
[mtch["comp_level"] for mtch in matches]

['f',
 'f',
 'qf',
 'qf',
 'qf',
 'qf',
 'qf',
 'qf',
 'qf',
 'qf',
 'qf',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'qm',
 'sf',
 'sf',
 'sf',
 'sf']

That's a long list. We just want to see what values the *comp_level* field can be. We don't need the entire list. Fortunately Python provides an easy way to extract only the unique items from a list.

In [73]:
# Unique competition levels
set([mtch["comp_level"] for mtch in matches])

{'f', 'qf', 'qm', 'sf'}

That's better. There are four types of matches in the dataset: qualification (qm), quarter-finals (qf), semi-finals (sf), and finals (f).

The built-in `set()` function converts its argument to a new composite data type that we have not previously discussed. Python sets are similar to lists, except they are not allowed to have duplicate values. Sets can be created witht the `set()` function or by placing the desired contents within curly braces.

In [74]:
# Creating a set
myset = {1, 2, 2, 3, 3, 3, 4, 4, 4, 4}
print(type(myset))
myset

<class 'set'>


{1, 2, 3, 4}

#### 2. List of Match Scores
Suppose we are interested only in the total points scored by the red and blue alliances for qualificatio matches. The following list comprehension will extract that information.

In [75]:
# List Comprehension for total scores
total_scores = [{"match_number": mtch["match_number"],
                 "totalPoints_blue": mtch["score_breakdown"]["blue"]["totalPoints"],
                 "totalPoints_red": mtch["score_breakdown"]["red"]["totalPoints"]}
                for mtch in matches if mtch["comp_level"] == "qm"]
total_scores[:10]

[{'match_number': 1, 'totalPoints_blue': 69, 'totalPoints_red': 22},
 {'match_number': 10, 'totalPoints_blue': 53, 'totalPoints_red': 65},
 {'match_number': 11, 'totalPoints_blue': 91, 'totalPoints_red': 41},
 {'match_number': 12, 'totalPoints_blue': 59, 'totalPoints_red': 57},
 {'match_number': 13, 'totalPoints_blue': 57, 'totalPoints_red': 59},
 {'match_number': 14, 'totalPoints_blue': 84, 'totalPoints_red': 10},
 {'match_number': 15, 'totalPoints_blue': 186, 'totalPoints_red': 44},
 {'match_number': 16, 'totalPoints_blue': 33, 'totalPoints_red': 86},
 {'match_number': 17, 'totalPoints_blue': 107, 'totalPoints_red': 61},
 {'match_number': 18, 'totalPoints_blue': 89, 'totalPoints_red': 77}]

See how we built a dictionary within the list comprehension?

#### 3. Using a Nested Comprehension
Suppose we just want a list of all teams that participated in the competition?

In [76]:
# Simple list comprehension to extract blue alliance teams
blue_alliances = [mtch["alliances"]["blue"]["team_keys"] for mtch in matches]
blue_alliances[:5]

[['frc2930', 'frc2976', 'frc4918'],
 ['frc2930', 'frc2976', 'frc4918'],
 ['frc4512', 'frc949', 'frc4131'],
 ['frc4512', 'frc949', 'frc4131'],
 ['frc1318', 'frc4513', 'frc7461']]

That's not bad, but it's still a list of lists, with a lot of duplication. We would prefer just a simple, flat list. A nested list comprehension can help us achieve that.

In [77]:
# Nested list comprehension
blue_teams_1 = [teams[0] for teams in [mtch["alliances"]["blue"]["team_keys"] for mtch in matches]]
blue_teams_1[:5]

['frc2930', 'frc2930', 'frc4512', 'frc4512', 'frc1318']

Can you see what we did here? We embedded a list comprehension within another list comprehension. The inner comprehension extracted the blue alliance teams into a nested list, and the outer comprehension extracted the first element of each inner list. Nesting comprehensions enables sophisticated transformations of data. But we can do better.

In [78]:
blue_teams = [team for teams in [mtch["alliances"]["blue"]["team_keys"] for mtch in matches] for team in teams]
blue_teams[:10]

['frc2930',
 'frc2976',
 'frc4918',
 'frc2930',
 'frc2976',
 'frc4918',
 'frc4512',
 'frc949',
 'frc4131',
 'frc4512']

The syntax is confusing, but it gave us exactly what we want. All teams have been extracted into a single flat list. There are now two `for` expressions in the outer comprehension. The syntax will make more sense if we compare the comprehension to the equivalent `for` loops.

In [79]:
blue_teams = []
for teams in [mtch["alliances"]["blue"]["team_keys"] for mtch in matches]:
    for team in teams:
        blue_teams.append(team)
              
blue_teams[:10]

['frc2930',
 'frc2976',
 'frc4918',
 'frc2930',
 'frc2976',
 'frc4918',
 'frc4512',
 'frc949',
 'frc4131',
 'frc4512']

The `for` expressions in the list comprehension are placed in the same order that we would use if we were writing traditional nested `for` statements.

## IV. Exercises
Several of the following exercises require you to create text or JSON files. Make a subfolder called *s10_files* and make sure your code saves the files to that folder.

### Ex 9.1
Create a string variable containing JSON text. The JSON string should contain a list of dictionary objects. Each dictionary object should represent a country of your choice. It should have keys and values for:
* The country's name
* The country's capitol
* Famous persons from that country. This value should contain a nested list with two or more famous person.

Pass the string variable to the `json.dumps()` method to verify the syntax is correct.

Remember, string values must be contained within double quotes. You can see the syntax rules for JSON at [json.org](http://json.org).

In [80]:
# Ex 9.1
import json
json_string = {"country": "", "country capital": "", "Famous persons": []}
json_string = json.dumps(json_string)
# Create JSON string here

# Leave this line alone.
# You will see an error if there is an error in your json string.
# Otherwise the json object will be displayed.
json.loads(json_string)

{'country': '', 'country capital': '', 'Famous persons': []}

### Ex 9.2
Use a list comprehension to create a list of tuples. Each tuple will have four items: 1) the *comp_level*, 2) the *match_number*, 3) the blue alliance score, and 4) the red alliance score. Save the list to a variable named `match_scores` and display the first six list elements.

In [81]:
# Ex 9.2
match_scores = [(mtch["comp_level"], mtch["match_number"], mtch["alliances"]["blue"]["score"],
                mtch["alliances"]["red"]["score"]) for mtch in matches]
match_scores[0:6]

[('f', 1, 136, 148),
 ('f', 2, 137, 163),
 ('qf', 1, 122, 193),
 ('qf', 2, 95, 140),
 ('qf', 1, 134, 89),
 ('qf', 2, 191, 128)]

### Ex 9.3
Save the list you created in **Ex IV.1** to a file using the JSON package. Name the file *match_scores.json*.

In [82]:
# Ex 9.3
with open("match_scores.json", "w") as scores:
    json.dump(match_scores, scores)

### Ex 9.4
Open the match_scores.json file using the json package. When opening the file, assign the contents to a variable called `match_scores2`. Display the last five elements in the `match_scores2` list.

In [83]:
# Ex 9.4
with open("match_scores.json", "r") as scores:
    match_scores2 = json.load(scores)
match_scores2[-5:]

[['qm', 9, 64, 113],
 ['sf', 1, 139, 171],
 ['sf', 2, 94, 137],
 ['sf', 1, 147, 130],
 ['sf', 2, 158, 91]]

### Ex 9.5
Save the `match_scores2` list you created in **EX IV.3** to a text file called *match_scores.txt* (do not use the JSON module). Each row of the file should contain one tuple. Don't forget that each row needs to end with a newline character.

In [118]:
# Ex 9.5
with open("match_scores.txt", "wt") as scores:
    for mtch in match_scores2:
        line = str(mtch) + "\n"
        scores.write(line)
        print(mtch, "\n")

['f', 1, 136, 148] 

['f', 2, 137, 163] 

['qf', 1, 122, 193] 

['qf', 2, 95, 140] 

['qf', 1, 134, 89] 

['qf', 2, 191, 128] 

['qf', 1, 130, 133] 

['qf', 2, 132, 68] 

['qf', 3, 79, 83] 

['qf', 1, 130, 95] 

['qf', 2, 137, 32] 

['qm', 1, 69, 22] 

['qm', 10, 53, 65] 

['qm', 11, 91, 41] 

['qm', 12, 59, 57] 

['qm', 13, 57, 59] 

['qm', 14, 84, 10] 

['qm', 15, 186, 44] 

['qm', 16, 33, 86] 

['qm', 17, 107, 61] 

['qm', 18, 89, 77] 

['qm', 19, 30, 57] 

['qm', 2, 45, 91] 

['qm', 20, 146, 39] 

['qm', 21, 93, 84] 

['qm', 22, 82, 74] 

['qm', 23, 64, 57] 

['qm', 24, 80, 110] 

['qm', 25, 71, 46] 

['qm', 26, 31, 220] 

['qm', 27, 82, 79] 

['qm', 28, 56, 105] 

['qm', 29, 49, 89] 

['qm', 3, 12, 65] 

['qm', 30, 49, 82] 

['qm', 31, 112, 107] 

['qm', 32, 100, 57] 

['qm', 33, 36, 112] 

['qm', 34, 19, 93] 

['qm', 35, 49, 29] 

['qm', 36, 80, 121] 

['qm', 37, 71, 118] 

['qm', 38, 97, 109] 

['qm', 39, 48, 145] 

['qm', 4, 84, 98] 

['qm', 40, 117, 67] 

['qm', 41, 69, 92] 



### Ex 9.6
Use the `match_scores2` list and a list comprehension to create a list of combined match scores (i.e., the sum of the blue alliance and red alliance scores). Don't worry about tracking the matches. Then use the built-in `max()` function to find the highest combined score for a match. Finally, use the built-in `min()` function to find the lowest combined score for a match.

In [85]:
# Ex 9.6m
mtch_both =[]
mtch_red = [mtch[2] for mtch in match_scores2]
mtch_blue = [mtch[3] for mtch in match_scores2]
for mtch in match_scores:
    mtch_both.append(int(str(mtch[2])) + int(str(mtch[3])))
print(max(mtch_both))
print(min(mtch_both))

319
61


### Ex 9.7
Open the *match_scores.txt* file. Iterate through each row of the file and find the match (i.e., competition level and match number) that had the highest combined score. Also find the match that had the lowest combined score. 

This one is tricky. Here are some hints.
* Use string indexing to stript the beginning and ending square brackets from each line.
* Use the `str.split()` method to split the string between the commas. See the [documentation for `str.split()` here](https://docs.python.org/3/library/stdtypes.html?highlight=split#str.split).
* Don't forget to convert data types!

In [237]:
# Ex 9.7
score = {}
with open("match_scores.txt", "rt") as mtch_scores:
    for matches in mtch_scores.readlines():
        mtch = matches[1:-2]
        mtch = mtch.split(",")
        comp_match = {mtch[1]:int(mtch[2]) + int(mtch[3])}
        score.update(comp_match)
    max_key = max(score, key=score.get)
    max_value = max(score.values())
    print(max_key, max_value)

 65 317


### Ex 9.8
The time values in the matches JSON data are all big integers, like *1583104716*. These are UNIX style time stamps. A UNIX timestamp is the number of seconds that have elapsed since 00:00 a.m. on January 1st, 1970. Use a list comprehension and methods from the `datetime` module in the standard library to build a list of match start times that is human readable.
* You will need to review the [documentation for the datetime module](https://docs.python.org/3/library/datetime.html).
* Check out the `datetime.fromtimestamp()` method.
* Also check out the `datetime.strftime()` method.
* It seems redudent, but the `datetime` module contains a `datetime` class. This is an important detail if you want your code to work.

In [292]:
# Ex 9.8
from datetime import datetime;
[mtch1.strftime("%B, %d, %Y, %X") for mtch1 in [datetime.fromtimestamp(match) for match in [mtch["actual_time"] for mtch in matches]]]

['March, 01, 2020, 15:18:36',
 'March, 01, 2020, 15:56:12',
 'March, 01, 2020, 13:03:52',
 'March, 01, 2020, 13:38:43',
 'March, 01, 2020, 13:13:35',
 'March, 01, 2020, 13:47:30',
 'March, 01, 2020, 13:20:22',
 'March, 01, 2020, 14:00:29',
 'March, 01, 2020, 14:20:27',
 'March, 01, 2020, 13:31:40',
 'March, 01, 2020, 14:07:20',
 'February, 29, 2020, 11:04:24',
 'February, 29, 2020, 12:39:12',
 'February, 29, 2020, 12:47:13',
 'February, 29, 2020, 12:56:47',
 'February, 29, 2020, 13:07:26',
 'February, 29, 2020, 13:17:49',
 'February, 29, 2020, 14:07:40',
 'February, 29, 2020, 14:15:20',
 'February, 29, 2020, 14:22:20',
 'February, 29, 2020, 14:29:26',
 'February, 29, 2020, 14:37:37',
 'February, 29, 2020, 11:16:52',
 'February, 29, 2020, 14:45:04',
 'February, 29, 2020, 14:53:53',
 'February, 29, 2020, 15:02:23',
 'February, 29, 2020, 15:09:56',
 'February, 29, 2020, 15:16:37',
 'February, 29, 2020, 15:23:21',
 'February, 29, 2020, 15:33:08',
 'February, 29, 2020, 15:42:19',
 'February

## V. Quiz
Enter your answers as comments in the code cells below each question.

**#1.** What is the absolute path of this notebook file on your computer?

In [293]:
#1 
# C:\Users\pperv\Desktop\Robotics\Robotics Training\pyclass_frc\sessions\s09_files_and_comprehensions\s09_files_and_comprehensions.ipynb

**#2.** What is the relative path of the session 2 notebook, *s02_data_types.ipynb* from this notebook?

In [294]:
#2
# ../s02_data_types/s02_data_types.ipynb

**#3.** In general, is it better to use absolute or relative paths in computer programs?

In [295]:
#3
# Relative paths as absolute paths will only allow specific users to use the computer programs.

**#4.** Why is it beneficial to use the `with` statement when opening or saving to files?

In [296]:
#4
# The with file makes sure to closes the file when the code is done running or when the code produces an error.

**#5.** What does putting an exclamation mark, `!`, at the beginning of a line in a code cell do?

In [1]:
#5
# We can run CLI commands from Jupyter Notebook using the !.

**#6.** What module must be imported in order to change the current working directory? What is the command for changing the current working directory?

In [2]:
#6
# You have to import the os module.The command to change the current working directory is os.chdir.

**#7.** Earlier in the notebook there was a cell that contained the statement `os.chdir("misc_files")`. The cell runs one time with no problems. But if you try to run it a second time, it throws an error. Why?

In [3]:
#7
# You can't change the directory to a directory you are currently working in.


**#8.** What will each of the following strings look like when printed? Type out what you thing the output will be in your answer.
* "1/n2\n3"
* "1\n2\\\\n3"
* "\\"1\\"\\\\m2"

In [299]:
#8
#1/n2
#3

#1
#2\n3

#"1"\m2

**#9.** The last code example in section 1.D demonstrated how to read a file line by line. Would it be a good idea to use this technique to read the *matches.json* file? Why or why not? (HINT: How many newline characters are in *matches.json*?)

In [4]:
#9
#It would not be a good idea because each line contains a dictionary item or a list value and then moves to the next line.

**#10.** What 1970's operating system greatly influencd MS-DOS, and subsequently Microsoft Windows?

In [5]:
#10
#CP/M

## VI. Save Your Work
Once you have completed the exercises, save a copy of the notebook and the two files you created (*match_scores.json* and *match_scores.txt*) outside of the git repository (outside of the *pyclass_frc* folder). Upload the files to your git assignments repository.

## VII. Concept and Terminology Review
You should be able to define the following terms or describe the concept.
* Built-in `open()` function
* File read and write modes
* Relative and absolute file paths
* Understanding Python's working directory
* Changing the working directory
* Closing files
* Using the `with` statement to work with files
* Newline characters
* Inserting newlines in strings
* Javascript Object Notation (JSON)
* Python's json package
* List and dictionary comprehensions
* Filtering within comprehensions
* Nested and multi-level comprehensions
* Python's built-in `min()` and `max()` functions
* Converting Unix timestamps with Python's datetime module
* The `str.split()` method.

[Table of Contents](../../index.ipynb)